rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,173 @@
+   their business logic just so they can access them within the context of a
+   handler.
+
+5. Many capabilities provided by services should not be arbitrarily used in
+   unconstrained execution such as in the plugin lifecycle hooks. For example,
+   it's appropriate for an Elasticsearch client to throw an error if it's used
+   inside an API route and Elasticsearch isn't available, however it's not
+   appropriate for a plugin to throw an error in their start function if
+   Elasticsearch is not available. If the ES client was only made available
+   within the handler context and not to the plugin's start contract at large,
+   then this isn't an issue we'll encounter in the first place.
+
+# Detailed design
+
+There are two parts to this proposal. The first is the handler interface
+itself, and the second is the interface that a service owner implements to make
+their handlers extensible.
+
+## Handler interface
+
+```ts
+interface HandlerContext {
+  core?: Record<string, unknown>;
+  [serviceName: string]: Record<string, unknown>;
+}
+
+type Handler = async (HandlerContext, ...args: unknown[]) => unknown;
+```
+
+## Handler creator interface
+
+```ts
+type CapabilityCreator = async (...args: unknown[]) => Record<string, unknown>;
+
+interface HandlerCreator {
+  registerCapability(serviceName: string, CapabilityCreator);
+}
+```
+
+Note: The `HandlerCreator` is defined as a literal interface in this document,
+but in practice this interface is just a guide for the pattern of registering
+capabilities. Certain services may have multiple different types of handlers,
+so they may choose not to use the generic name `registerCapability` in favor of
+something more explicit.
+
+## End to end example
+
+```js
+http.router.registerRequestCapability('elasticsearch', async request => {
+  const client = await core.elasticsearch.client$.toPromise();
+  return client.child({
+    headers: { authorization: request.headers.authorization },
+  });
+});
+
+http.router.route({
+  path: '/foo',
+  async routeHandler(context) {
+    context.elasticsearch.search(); // === callWithRequest(request, 'search')
+  },
+});
+```
+
+# Drawbacks
+
+- Since the service names are passed as strings but accessed as properties,
+  it's unclear how best to type these context objects. We can expose types at
+  the core and plugin level, but consumers of those types might need to compose
+  them together manually.
+- The longer we wait to implement this, the more refactoring of newer code
+  we'll need to do to roll this out.
+- It's a new formal concept and set of terminology that developers will need to
+  learn relative to other new platform terminology.
+- Handlers are a common pattern for HTTP route handlers, but people don't
+  necessarily associate similar patterns elsewhere as the same set of problems.
+- "Chicken and egg" questions will arise around where capabilities should be
+  registered. For example, does the `http` service invoke its
```

Comment:
it's even worse for x-pack capabilities, such as security, where core cannot relay on capability existence (or it can be disabled, as any plugin)